src/dbnode/storage/database.go

```diff
@@ -533,20 +555,29 @@ func (d *db) ShardSet() sharding.ShardSet {
 	return shardSet
 }
 
-func (d *db) queueBootstrapWithLock() {
+func (d *db) enqueueBootstrap(onCompleteFn func()) {
 	// Only perform a bootstrap if at least one bootstrap has already occurred. This enables
 	// the ability to open the clustered database and assign shardsets to the non-clustered
 	// database when it receives an initial topology (as well as topology changes) without
 	// triggering a bootstrap until an external call initiates a bootstrap with an initial
 	// call to Bootstrap(). After that initial bootstrap, the clustered database will keep
 	// the non-clustered database bootstrapped by assigning it shardsets which will trigger new
 	// bootstraps since d.bootstraps > 0 will be true.
-	if d.bootstraps > 0 {
+	d.RLock()
+	shouldBootstrap := d.bootstraps > 0
+	d.RUnlock()
+	if shouldBootstrap {
```

Comment:
Any chances of enqueueing 2 competing bootstraps in case of adding 2 new namespaces at once (or in a quick succession)?

Reply:
Competing bootstraps is already handled in `bootstrapManager.startBootstrap()`. We can check if bootstrap is already started here by using `bootstrapAsyncResult.bootstrapResult.AlreadyBootstrapping`. I think I'll need to update this method so that we need to invoke`onCompleteFn` only if we are not already bootstrapping because otherwise we might accidentally enableFileOps when another bootstrap is still in progress. Thanks for pointing this out ;)